File: Project-Tech/Backend/unlockit/src/main/java/io/unlockit/google/GoogleUtils.java
package io.unlockit.google;

import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class GoogleUtils {

    private static final String TOKEN_INFO_URL = "https://www.googleapis.com/oauth2/v1/tokeninfo?access_token=";

    private static boolean isTokenValid(String accessToken) {
        String url = TOKEN_INFO_URL + accessToken;
        boolean isValid = false;
        try {
            HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
            connection.setRequestMethod("GET");
            int responseCode = connection.getResponseCode();
            if (responseCode == HttpURLConnection.HTTP_OK) {
                isValid = true;
                // Uncomment to view token details
                /*
                try (InputStream inputStream = connection.getInputStream();
                     Scanner scanner = new Scanner(inputStream)) {

                    String response = scanner.useDelimiter("\\A").next();
                    System.out.println("response: " + response);
                }
                 */
            }
        }
        catch (IOException e) {
            System.out.println("Failed to connect to " + TOKEN_INFO_URL);
        }
        return isValid;
    }

    private static String extractAccessToken(String authorizationHeader) {
        if (authorizationHeader != null && authorizationHeader.startsWith("Bearer ")) {
            return authorizationHeader.substring(7);
        }
        return null;
    }

    public static boolean isRequestAuthorized(String authorizationHeader) {
        String accessToken = extractAccessToken(authorizationHeader);
        if (accessToken == null) {
            return false;
        }
        //System.out.println("Access Token: " + accessToken);
        if (!isTokenValid(accessToken)) {
            return false;
        }
        return true;
    }

    public static List<String> getUserInfo(String authorizationHeader) {
        String accessToken = extractAccessToken(authorizationHeader);
        return getUserInfoList(accessToken);
    }

    public static String getUserLoginExpireTime(String authorizationHeader) {
        String accessToken = extractAccessToken(authorizationHeader);
        return getLoginExpireTime(accessToken);
    }

    private static String getLoginExpireTime(String accessToken) {
        String url = TOKEN_INFO_URL + accessToken;
        try {
            HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
            connection.setRequestMethod("GET");
            int responseCode = connection.getResponseCode();
            if (responseCode == HttpURLConnection.HTTP_OK) {
                try (InputStream inputStream = connection.getInputStream();
                     Scanner scanner = new Scanner(inputStream)) {

                    String response = scanner.useDelimiter("\\A").next();
                    //System.out.println("response: " + response);

                    // Parse the response to extract userId and userEmail
                    JSONObject jsonObject = new JSONObject(response);
                    String expireTime = String.valueOf(jsonObject.get("expires_in"));
                    return expireTime;
                }
            }
        }
        catch (IOException e) {
            System.out.println("Failed to connect to " + TOKEN_INFO_URL);
        }
        return null;
    }

    private static List<String> getUserInfoList(String accessToken) {
        String url = TOKEN_INFO_URL + accessToken;
        List<String> userInfo = new ArrayList<>();
        try {
            HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
            connection.setRequestMethod("GET");
            int responseCode = connection.getResponseCode();
            if (responseCode == HttpURLConnection.HTTP_OK) {
                try (InputStream inputStream = connection.getInputStream();
                     Scanner scanner = new Scanner(inputStream)) {

                    String response = scanner.useDelimiter("\\A").next();
                    //System.out.println("response: " + response);

                    // Parse the response to extract userId and userEmail
                    JSONObject jsonObject = new JSONObject(response);
                    String userId = jsonObject.getString("user_id");
                    String userEmail = jsonObject.getString("email");

                    userInfo.add(userId);
                    userInfo.add(userEmail);
                    return userInfo;
                }
            }
        }
        catch (IOException e) {
            System.out.println("Failed to connect to " + TOKEN_INFO_URL);
        }
        return null;
    }
}

File: Project-Tech/Hyperledger-Fabric/client/src/google/googleUtils.ts
import { get } from 'https';

class GoogleUtils {
    private static readonly TOKEN_INFO_URL = "https://www.googleapis.com/oauth2/v1/tokeninfo?access_token=";

    private static isTokenValid(accessToken: string): Promise<boolean> {
        const url = GoogleUtils.TOKEN_INFO_URL + accessToken;

        return new Promise<boolean>((resolve, reject) => {
            get(url, (response) => {
                if (response.statusCode === 200) {
                    resolve(true);
                } else {
                    resolve(false);
                }
            }).on('error', (error) => {
                console.error("Failed to connect to " + GoogleUtils.TOKEN_INFO_URL);
                reject(error);
            });
        });
    }

    private static extractAccessToken(authorizationHeader: string | undefined): string | null {
        if (authorizationHeader !== null && authorizationHeader.startsWith("Bearer ")) {
            return authorizationHeader.substring(7);
        }
        return null;
    }

    public static async isRequestAuthorized(authorizationHeader: string | undefined): Promise<boolean> {
        const accessToken = GoogleUtils.extractAccessToken(authorizationHeader);
        if (!accessToken) {
            return false;
        }

        try {
            const isValid = await GoogleUtils.isTokenValid(accessToken);
            return isValid;
        } catch (error) {
            return false;
        }
    }

    public static async getUserInfo(authorizationHeader: string | undefined): Promise<string[] | null> {
        const accessToken = GoogleUtils.extractAccessToken(authorizationHeader);
        if (!accessToken) {
            return null;
        }

        return GoogleUtils.getUserInfoList(accessToken);
    }

    public static async getUserLoginExpireTime(authorizationHeader: string | undefined): Promise<string | null> {
        const accessToken = GoogleUtils.extractAccessToken(authorizationHeader);
        if (!accessToken) {
            return null;
        }

        return GoogleUtils.getLoginExpireTime(accessToken);
    }

    private static async getLoginExpireTime(accessToken: string): Promise<string | null> {
        const url = GoogleUtils.TOKEN_INFO_URL + accessToken;

        try {
            const response = await new Promise<string>((resolve, reject) => {
                get(url, (response) => {
                    let data = '';
                    response.on('data', (chunk) => {
                        data += chunk;
                    });

                    response.on('end', () => {
                        resolve(data);
                    });
                }).on('error', (error) => {
                    console.error("Failed to connect to " + GoogleUtils.TOKEN_INFO_URL);
                    reject(error);
                });
            });

            const jsonObject = JSON.parse(response);
            const expireTime = jsonObject.expires_in;
            return String(expireTime);
        } catch (error) {
            return null;
        }
    }

    private static async getUserInfoList(accessToken: string): Promise<string[] | null> {
        const url = GoogleUtils.TOKEN_INFO_URL + accessToken;

        try {
            const response = await new Promise<string>((resolve, reject) => {
                get(url, (response) => {
                    let data = '';
                    response.on('data', (chunk) => {
                        data += chunk;
                    });

                    response.on('end', () => {
                        resolve(data);
                    });
                }).on('error', (error) => {
                    console.error("Failed to connect to " + GoogleUtils.TOKEN_INFO_URL);
                    reject(error);
                });
            });

            const jsonObject = JSON.parse(response);
            const userId = jsonObject.user_id;
            const userEmail = jsonObject.email;

            return [userId, userEmail];
        } catch (error) {
            return null;
        }
    }
}

export default GoogleUtils;
